src/main/perf/DocGrouper.java

```diff
@@ -0,0 +1,438 @@
+                groupIds = group1M;
+            } else if (numDocs >= 500000) {
+                groupIds = group100K;
+            } else {
+                groupIds = group10K;
+            }
+            docsPerGroupBlock = ((float) numDocs) / groupIds.length;
+            reset();
+        }
+
+        @Override
+        public void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException {
+            if (lfd == LineFileDocs.END) {
+                assert docCounter == 0;
+                outputQueue.put(END);
+            }
+            buffer[docCounter++] = lfd;
+            if (docCounter == nextNumDocs) {
+                outputQueue.put(new DocGroups.TextBased(groupIds[groupCounter], buffer));
+                groupCounter++;
+                reset();
+            }
+        }
+
+        /* Called when we move to next group */
+        private void reset() {
+            nextNumDocs = calculateNextGroupDocNum();
+            buffer = new LineFileDocs.LineFileDoc[nextNumDocs];
+        }
+
+        private int calculateNextGroupDocNum() {
+            // This will toggle between X and X+1 docs,
+            // converging over time on average to the
+            // floating point docsPerGroupBlock:
+            if (groupCounter == groupIds.length - 1) {
+                return numDocs - ((int) (groupCounter * docsPerGroupBlock));
+            } else {
+                return ((int) ((1 + groupCounter) * docsPerGroupBlock)) - ((int) (groupCounter * docsPerGroupBlock));
+            }
+        }
+    }
+
+    /**
+     * The binary LFD is naturally grouped by the binary blob, so that if we have two groups sharing the same
+     * binary blob then the two groups cannot be indexed concurrently.
+     * This grouper will produce {@link DocGroups.BinaryBased}, which will contain either:
+     * 1. One group, consists of multiple LFD, or
+     * 2. One LFD, split into multiple groups
+     * Such that each {@link DocGroups.BinaryBased} can be indexed in parallel
+     * <br>
+     * The algorithm will first calculate average number of documents in the group, then init first budget as the avg
+     * value, then tries to follow:
+     * 1. If the coming LFD has number of documents within 0.5 * budget and 1.5 * budget, then the LFD sole will form
+     *    a group, we will then adjust next budget as: newBudget = 2 * budget - LFD.docNum()
+     * 2. If the coming LFD has number of documents less than 0.5 * budget, then we will try to accumulate the next LFDs
+     *    until the total number of documents of the accumulated LFD reaches 0.5 * budget and form the group.
+     *    The next budget will be adjusted as: newBudget = 2 * budget - total_number_doc_of_the_group
+     * 3. If the coming LFD has number of documents larger than 1.5 * budget, then we will evenly divide this LFD into
+     *    round(LFD.docNum() / budget) number of groups. Then adjust the next budget as:
+     *    newBudget = (num_group + 1) * budget - LFD.docNum()
```

Comment:
OMG so hairy!

I wonder if we should simply disallow grouping when indexing from the binary line docs file?

The binary format was created to do our best to absolutely minimize the doc processing overhead during indexing, so that our reported throughput in the nightly benchy is as close as possible to Lucene's raw indexing speed and not measuring the cost of doc preparation.  I double checked the nightly benchy and when it measures its "fast" indexing (building a non-deterministic index that will not be used for searching) it is indeed not testing the performance of indexing groups.  So this means we could fix the nightly benchy to use the text line file docs when indexing groups?

Reply:
Oh that's a great great news! As I guess binary LFD is only used by nightly bench, we probably can deprecate binary format totally?
I have a PR for it, let me know what you think: https://github.com/mikemccand/luceneutil/pull/228